Support 'termination protection' for cloudformation stacks

prasadkatti opened the pull request:
##### SUMMARY
Add support for Cloudformation stack termination protection

Fixes #31462

https://aws.amazon.com/blogs/mt/use-aws-cloudformation-stack-termination-protection-and-rollback-triggers-to-maintain-infrastructure-availability/
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
`cloudformation`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
  -->
This PR adds support for enabling/disabling `Termination Protection` for cloudformation stacks.
It adds a new option called `termination_protection`, which is a boolean value.

Termination protection can be enabled during stack creation for new stacks. For existing stacks, the option can be used to enable/disable termination protection. When this option is used for existing stacks, the task will only be used to change this particular setting. 

Stack deletion tasks do not read this option. That is, you can't disable the termination protection and delete the stack in the same task.

prasadkatti committed:
Support 'termination protection' for cloudformation stacks

prasadkatti committed:
Fix failure msg value

tedder commented on lib/ansible/modules/cloud/amazon/cloudformation.py:
```diff
@@ -197,6 +201,16 @@
       Stack: ansible-cloudformation
+
+# Enable termination protection on a stack.
+# If the stack already exists, this will update it's termination protection
```
it's -> its

tedder commented on lib/ansible/modules/cloud/amazon/cloudformation.py:
```diff
@@ -280,8 +294,11 @@ def create_stack(module, stack_params, cfn):
+    # apply on creation, not update.
     stack_params['DisableRollback'] = module.params['disable_rollback']
+    stack_params['EnableTerminationProtection'] = bool(module.params.get(
+        'termination_protection'))
```
this line break is .. awful. Can you find another pep8 way to do it? maybe start the second line with 'module'

prasadkatti replied:
looks like lines can be as long as 160 chars. So removed some unnecessary line-breaks.

tedder commented on lib/ansible/modules/cloud/amazon/cloudformation.py:
```diff
@@ -351,6 +368,24 @@ def update_stack(module, stack_params, cfn):
+    '''updates termination protection of a stack'''
+    stack = get_stack_facts(cfn, stack_params['StackName'])
+    if stack:
+        if stack['EnableTerminationProtection'] is not module.params['termination_protection']:
```
I think, unless you have a good reason, that you should use stack_params here and four lines down instead of module.params. Honestly I wish we didn't have to push module around.

prasadkatti replied:
Passing in stack_name and the desired state of 'termination protection' instead of looking into module.params inside the function.
Also, since I am using `exit_json` and `fail_json`, I need `module`.

tedder replied:
Yeah, I know you need module. That was just a general musing.

prasadkatti committed:
Avoid using module.params inside update_termination_protection

- Pass in the stack_name and desired termination protection state to update_termination_protection
  and avoid using module.params
- Removed some unnecessary line breaks

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/40744/2/tests

prasadkatti committed:
Fix for failing cloudformation unit test

tedder commented:
`shipit`

ryansb commented on lib/ansible/modules/cloud/amazon/cloudformation.py:
```diff
@@ -536,6 +570,7 @@ def main():
     cfn.describe_stacks = backoff_wrapper(cfn.describe_stacks)
     cfn.list_stack_resources = backoff_wrapper(cfn.list_stack_resources)
     cfn.delete_stack = backoff_wrapper(cfn.delete_stack)
+    cfn.update_termination_protection = backoff_wrapper(cfn.update_termination_protection)
```
We can't just assume people have the latest boto3, so when using newer features we have to add a check for the function being available, such as `if hasattr(cfn, 'update_termination_protection')` and then fail if that feature is missing and the user specified `termination_protection`. If they didn't set `termination_protection` and are on an older `boto3` they still should be able to use the module.

prasadkatti replied:
agreed. I will add that in. Thanks for the review.

prasadkatti committed:
Check if cfn has update_termination_protection attr

prasadkatti committed:
Use hasattr to test if cfn supports update_termination_protection

prasadkatti commented:
@ansibot ready_for_review

prasadkatti commented:
bot_status

prasadkatti commented:
ready_for_review

prasadkatti commented:
@ansibot says this is waiting on me. How do i use "ready_for_review"? Do I need to reply to the review comment?

ryansb commented:
@prasadkatti you're all set, just needs me & one other person to review

LGTM 

cc/ @tedder

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudformation.py:
```diff
@@ -554,6 +601,8 @@ def main():
             result = create_stack(module, stack_params, cfn)
         elif module.params.get('create_changeset'):
             result = create_changeset(module, stack_params, cfn)
+        elif module.params.get('termination_protection') is not None:
```
This is working almost as expected for me. It seems like now general stack updates will not occur unless termination_protection is None.

I'm wondering if update_stack() should call update_termination_protection().

That would also allow the exit output to stay the same if no update is being made - with this change if I call the module twice with termination_protection: false then I get the message:
`"msg": "Termination protection on stack ... is already the requested value"` whereas normally if a stack has no updates we would get the output: `"output": "Stack is already up-to-date."`.

prasadkatti replied:
Oh, good catch and thanks for the review. This will actually break one of my own playbooks as I create a stack and register the output for use in subsequent tasks. I will make the required changes and let you know. 👍

prasadkatti committed:
termination_protection shouldn't prevent update_stack call for existing stacks

prasadkatti commented:
ready_for_review

s-hertel commented:
Thanks @prasadkatti!

jonathannaguin commented:
@s-hertel @ryansb can this be back ported to the 2.4.x branch? Otherwise it seems we need to wait until Feb 2018 to get this feature included (correct me if I am wrong).

s-hertel commented:
@jonathannaguin We don't backport features (just bug fixes/security fixes). You can copy this file into your local library/ to override the 2.4 version of the module.
